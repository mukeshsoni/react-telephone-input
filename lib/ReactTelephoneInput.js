'use strict';

// TODO - fix the onlyContries props. Currently expects that as an array of country object, but users should be able to send in array of country isos

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var some = require('lodash/some');
var find = require('lodash/find');
var reduce = require('lodash/reduce');
var map = require('lodash/map');
var filter = require('lodash/filter');
var findIndex = require('lodash/findIndex');
var first = require('lodash/first');
var tail = require('lodash/tail');
var debounce = require('lodash/debounce');
var memoize = require('lodash/memoize');
var assign = require('lodash/assign');
var isEqual = require('lodash/isEqual');
// import lodash string methods
var trim = require('lodash/trim');
var startsWith = require('lodash/startsWith');

var React = require('react');
var ReactDOM = require('react-dom');
var onClickOutside = require('react-onclickoutside');
var classNames = require('classnames');
var countryData = require('country-telephone-data');
var allCountries = countryData.allCountries;
var iso2Lookup = countryData.iso2Lookup;
var allCountryCodes = countryData.allCountryCodes;

if (typeof document !== 'undefined') {
    var isModernBrowser = Boolean(document.createElement('input').setSelectionRange);
} else {
    var isModernBrowser = true;
}

var keys = {
    UP: 38,
    DOWN: 40,
    RIGHT: 39,
    LEFT: 37,
    ENTER: 13,
    ESC: 27,
    PLUS: 43,
    A: 65,
    Z: 90,
    SPACE: 32
};

function isNumberValid(inputNumber) {
    var countries = countryData.allCountries;
    return some(countries, function (country) {
        return startsWith(inputNumber, country.dialCode) || startsWith(country.dialCode, inputNumber);
    });
}

var ReactTelephoneInput = exports.ReactTelephoneInput = React.createClass({
    displayName: 'ReactTelephoneInput',
    getInitialState: function getInitialState() {
        var preferredCountries = this.props.preferredCountries.map(function (iso2) {
            return iso2Lookup.hasOwnProperty(iso2) ? allCountries[iso2Lookup[iso2]] : null;
        }).filter(function (val) {
            return val !== null;
        });

        return assign({}, {
            preferredCountries: preferredCountries,
            showDropDown: false,
            queryString: '',
            freezeSelection: false,
            debouncedQueryStingSearcher: debounce(this.searchCountry, 300)
        }, this._mapPropsToState(this.props, true));
    },

    propTypes: {
        value: React.PropTypes.string,
        initialValue: React.PropTypes.string,
        autoFormat: React.PropTypes.bool,
        defaultCountry: React.PropTypes.string,
        onlyCountries: React.PropTypes.arrayOf(React.PropTypes.object),
        preferredCountries: React.PropTypes.arrayOf(React.PropTypes.string),
        classNames: React.PropTypes.string,
        className: React.PropTypes.string,
        inputId: React.PropTypes.string,
        onChange: React.PropTypes.func,
        onEnterKeyPress: React.PropTypes.func,
        onBlur: React.PropTypes.func,
        onFocus: React.PropTypes.func,
        disabled: React.PropTypes.bool,
        pattern: React.PropTypes.string,
        required: React.PropTypes.bool
    },
    getDefaultProps: function getDefaultProps() {
        return {
            autoFormat: true,
            onlyCountries: allCountries,
            defaultCountry: allCountries[0].iso2,
            isValid: isNumberValid,
            flagsImagePath: 'flags.png',
            onEnterKeyPress: function onEnterKeyPress() {},
            preferredCountries: [],
            disabled: false,
            placeholder: '+1 (702) 123-4567',
            autoComplete: 'tel',
            required: false
        };
    },
    getNumber: function getNumber() {
        return this.state.formattedNumber !== '+' ? this.state.formattedNumber : '';
    },
    getValue: function getValue() {
        return this.getNumber();
    },
    componentDidMount: function componentDidMount() {
        document.addEventListener('keydown', this.handleKeydown);

        this._cursorToEnd(true);
        if (typeof this.props.onChange === 'function') {
            this.props.onChange(this.state.formattedNumber, this.state.selectedCountry);
        }
    },
    shouldComponentUpdate: function shouldComponentUpdate(nextProps, nextState) {
        return !isEqual(nextProps, this.props) || !isEqual(nextState, this.state);
    },
    componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
        this.setState(this._mapPropsToState(nextProps));
    },
    componentWillUnmount: function componentWillUnmount() {
        document.removeEventListener('keydown', this.handleKeydown);
    },
    scrollTo: function scrollTo(country, middle) {
        if (!country) {
            return;
        }

        var container = ReactDOM.findDOMNode(this.refs.flagDropdownList);

        if (!container) {
            return;
        }

        var containerHeight = container.offsetHeight;
        var containerOffset = container.getBoundingClientRect();
        var containerTop = containerOffset.top + document.body.scrollTop;
        var containerBottom = containerTop + containerHeight;

        var element = country;
        var elementOffset = element.getBoundingClientRect();

        var elementHeight = element.offsetHeight;
        var elementTop = elementOffset.top + document.body.scrollTop;
        var elementBottom = elementTop + elementHeight;
        var newScrollTop = elementTop - containerTop + container.scrollTop;
        var middleOffset = containerHeight / 2 - elementHeight / 2;

        if (elementTop < containerTop) {
            // scroll up
            if (middle) {
                newScrollTop -= middleOffset;
            }
            container.scrollTop = newScrollTop;
        } else if (elementBottom > containerBottom) {
            // scroll down
            if (middle) {
                newScrollTop += middleOffset;
            }
            var heightDifference = containerHeight - elementHeight;
            container.scrollTop = newScrollTop - heightDifference;
        }
    },
    formatNumber: function formatNumber(text, pattern) {
        if (!text || text.length === 0) {
            return '+';
        }

        // for all strings with length less than 3, just return it (1, 2 etc.)
        // also return the same text if the selected country has no fixed format
        if (text && text.length < 2 || !pattern || !this.props.autoFormat) {
            return '+' + text;
        }

        var formattedObject = reduce(pattern, function (acc, character) {
            if (acc.remainingText.length === 0) {
                return acc;
            }

            if (character !== '.') {
                return {
                    formattedText: acc.formattedText + character,
                    remainingText: acc.remainingText
                };
            }

            return {
                formattedText: acc.formattedText + first(acc.remainingText),
                remainingText: tail(acc.remainingText)
            };
        }, { formattedText: '', remainingText: text.split('') });
        return formattedObject.formattedText + formattedObject.remainingText.join('');
    },


    // put the cursor to the end of the input (usually after a focus event)
    _cursorToEnd: function _cursorToEnd(skipFocus) {
        var input = this.refs.numberInput;
        if (skipFocus) {
            this._fillDialCode();
        } else {
            input.focus();

            if (isModernBrowser) {
                var len = input.value.length;
                input.setSelectionRange(len, len);
            }
        }
    },

    // memoize results based on the first 5/6 characters. That is all that matters
    guessSelectedCountry: function guessSelectedCountry(inputNumber) {
        var secondBestGuess = find(allCountries, { iso2: this.props.defaultCountry }) || this.props.onlyCountries[0];
        var inputNumberForCountries = inputNumber.substr(0, 4);
        if (trim(inputNumber) !== '') {
            var bestGuess = reduce(this.props.onlyCountries, function (selectedCountry, country) {

                // if the country dialCode exists WITH area code

                if (allCountryCodes[inputNumberForCountries] && allCountryCodes[inputNumberForCountries][0] === country.iso2) {
                    return country;

                    // if the selected country dialCode is there with the area code
                } else if (allCountryCodes[inputNumberForCountries] && allCountryCodes[inputNumberForCountries][0] === selectedCountry.iso2) {
                    return selectedCountry;

                    // else do the original if statement
                } else {
                    if (startsWith(inputNumber, country.dialCode)) {
                        if (country.dialCode.length > selectedCountry.dialCode.length) {
                            return country;
                        }
                        if (country.dialCode.length === selectedCountry.dialCode.length && country.priority < selectedCountry.priority) {
                            return country;
                        }
                    }
                }
                return selectedCountry;
            }, { dialCode: '', priority: 10001 }, this);
        } else {
            return secondBestGuess;
        }

        if (!bestGuess.name) {
            return secondBestGuess;
        }

        return bestGuess;
    },
    getElement: function getElement(index) {
        return ReactDOM.findDOMNode(this.refs['flag_no_' + index]);
    },
    handleFlagDropdownClick: function handleFlagDropdownClick() {
        var _this = this;

        if (this.props.disabled) {
            return;
        }
        // need to put the highlight on the current selected country if the dropdown is going to open up
        this.setState({
            showDropDown: !this.state.showDropDown,
            highlightCountry: find(this.props.onlyCountries, this.state.selectedCountry),
            highlightCountryIndex: findIndex(this.state.preferredCountries.concat(this.props.onlyCountries), this.state.selectedCountry)
        }, function () {
            // only need to scrool if the dropdown list is alive
            if (_this.state.showDropDown) {
                _this.scrollTo(_this.getElement(_this.state.highlightCountryIndex + _this.state.preferredCountries.length));
            }
        });
    },
    handleInput: function handleInput(event) {
        var formattedNumber = '+',
            newSelectedCountry = this.state.selectedCountry,
            freezeSelection = this.state.freezeSelection;

        // if the input is the same as before, must be some special key like enter etc.
        if (event.target.value === this.state.formattedNumber) {
            return;
        }

        // ie hack
        if (event.preventDefault) {
            event.preventDefault();
        } else {
            event.returnValue = false;
        }

        if (event.target.value.length > 0) {
            // before entering the number in new format, lets check if the dial code now matches some other country
            var inputNumber = event.target.value.replace(/\D/g, '');

            // we don't need to send the whole number to guess the country... only the first 6 characters are enough
            // the guess country function can then use memoization much more effectively since the set of input it gets has drastically reduced
            if (!this.state.freezeSelection || this.state.selectedCountry.dialCode.length > inputNumber.length) {
                newSelectedCountry = this.guessSelectedCountry(inputNumber.substring(0, 6));
                freezeSelection = false;
            }
            // let us remove all non numerals from the input
            formattedNumber = this.formatNumber(inputNumber, newSelectedCountry.format);
        }

        var caretPosition = event.target.selectionStart;
        var oldFormattedText = this.state.formattedNumber;
        var diff = formattedNumber.length - oldFormattedText.length;

        this.setState({
            formattedNumber: formattedNumber,
            freezeSelection: freezeSelection,
            selectedCountry: newSelectedCountry.dialCode.length > 0 ? newSelectedCountry : this.state.selectedCountry
        }, function () {
            if (isModernBrowser) {
                if (caretPosition === 1 && formattedNumber.length === 2) {
                    caretPosition++;
                }

                if (diff > 0) {
                    caretPosition = caretPosition - diff;
                }

                if (caretPosition > 0 && oldFormattedText.length >= formattedNumber.length) {
                    this.refs.numberInput.setSelectionRange(caretPosition, caretPosition);
                }
            }

            if (this.props.onChange) {
                this.props.onChange(this.state.formattedNumber, this.state.selectedCountry);
            }
        });
    },
    handleInputClick: function handleInputClick() {
        this.setState({ showDropDown: false });
    },
    handleFlagItemClick: function handleFlagItemClick(country) {
        var currentSelectedCountry = this.state.selectedCountry;
        var nextSelectedCountry = find(this.props.onlyCountries, country);

        // tiny optimization
        if (currentSelectedCountry.iso2 !== nextSelectedCountry.iso2) {
            var dialCodeRegex = RegExp('^(\\+' + currentSelectedCountry.dialCode + ')|\\+');
            var newNumber = this.state.formattedNumber.replace(dialCodeRegex, '+' + nextSelectedCountry.dialCode);
            var formattedNumber = this.formatNumber(newNumber.replace(/\D/g, ''), nextSelectedCountry.format);

            this.setState({
                showDropDown: false,
                selectedCountry: nextSelectedCountry,
                freezeSelection: true,
                formattedNumber: formattedNumber
            }, function () {
                this._cursorToEnd();
                if (this.props.onChange) {
                    this.props.onChange(formattedNumber, nextSelectedCountry);
                }
            });
        } else {
            this.setState({ showDropDown: false });
        }
    },
    handleInputFocus: function handleInputFocus(evt) {
        // trigger parent component's onFocus handler
        if (typeof this.props.onFocus === 'function') {
            this.props.onFocus(evt, this.state.formattedNumer, this.state.selectedCountry);
        }

        this._fillDialCode();
    },
    _mapPropsToState: function _mapPropsToState(props) {
        var firstCall = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

        var inputNumber = void 0;

        if (props.value) {
            inputNumber = props.value;
        } else if (props.initialValue && firstCall) {
            inputNumber = props.initialValue;
        } else if (this.state && this.state.formattedNumber && this.state.formattedNumber.length > 0) {
            inputNumber = this.state.formattedNumber;
        } else {
            inputNumber = '';
        }

        var selectedCountryGuess = this.guessSelectedCountry(inputNumber.replace(/\D/g, ''));
        var selectedCountryGuessIndex = findIndex(allCountries, selectedCountryGuess);
        var formattedNumber = this.formatNumber(inputNumber.replace(/\D/g, ''), selectedCountryGuess ? selectedCountryGuess.format : null);

        return {
            selectedCountry: selectedCountryGuess,
            highlightCountryIndex: selectedCountryGuessIndex,
            formattedNumber: formattedNumber
        };
    },
    _fillDialCode: function _fillDialCode() {
        // if the input is blank, insert dial code of the selected country
        if (this.refs.numberInput.value === '+') {
            this.setState({ formattedNumber: '+' + this.state.selectedCountry.dialCode });
        }
    },
    _getHighlightCountryIndex: function _getHighlightCountryIndex(direction) {
        // had to write own function because underscore does not have findIndex. lodash has it
        var highlightCountryIndex = this.state.highlightCountryIndex + direction;

        if (highlightCountryIndex < 0 || highlightCountryIndex >= this.props.onlyCountries.length + this.state.preferredCountries.length) {
            return highlightCountryIndex - direction;
        }

        return highlightCountryIndex;
    },

    // memoize search results... caching all the way
    _searchCountry: memoize(function (queryString) {
        if (!queryString || queryString.length === 0) {
            return null;
        }
        // don't include the preferred countries in search
        var probableCountries = filter(this.props.onlyCountries, function (country) {
            return startsWith(country.name.toLowerCase(), queryString.toLowerCase());
        }, this);
        return probableCountries[0];
    }),
    searchCountry: function searchCountry() {
        var probableCandidate = this._searchCountry(this.state.queryString) || this.props.onlyCountries[0];
        var probableCandidateIndex = findIndex(this.props.onlyCountries, probableCandidate) + this.state.preferredCountries.length;
        this.scrollTo(this.getElement(probableCandidateIndex), true);

        this.setState({
            queryString: '',
            highlightCountryIndex: probableCandidateIndex
        });
    },
    handleKeydown: function handleKeydown(event) {
        if (!this.state.showDropDown) {
            return;
        }

        // ie hack
        if (event.preventDefault) {
            event.preventDefault();
        } else {
            event.returnValue = false;
        }

        var self = this;
        function _moveHighlight(direction) {
            self.setState({
                highlightCountryIndex: self._getHighlightCountryIndex(direction)
            }, function () {
                self.scrollTo(self.getElement(self.state.highlightCountryIndex), true);
            });
        }

        switch (event.which) {
            case keys.DOWN:
                _moveHighlight(1);
                break;
            case keys.UP:
                _moveHighlight(-1);
                break;
            case keys.ENTER:
                this.handleFlagItemClick(this.state.preferredCountries.concat(this.props.onlyCountries)[this.state.highlightCountryIndex], event);
                break;
            case keys.ESC:
                this.setState({ showDropDown: false }, this._cursorToEnd);
                break;
            default:
                if (event.which >= keys.A && event.which <= keys.Z || event.which === keys.SPACE) {
                    this.setState({ queryString: this.state.queryString + String.fromCharCode(event.which) }, this.state.debouncedQueryStingSearcher);
                }
        }
    },
    handleInputKeyDown: function handleInputKeyDown(event) {
        if (event.which === keys.ENTER) {
            this.props.onEnterKeyPress(event);
        }
    },
    handleClickOutside: function handleClickOutside() {
        if (this.state.showDropDown) {
            this.setState({
                showDropDown: false
            });
        }
    },
    getCountryDropDownList: function getCountryDropDownList() {
        var self = this;
        var countryDropDownList = map(this.state.preferredCountries.concat(this.props.onlyCountries), function (country, index) {
            var itemClasses = classNames({
                country: true,
                preferred: findIndex(self.state.preferredCountries, { iso2: country.iso2 }) >= 0,
                highlight: self.state.highlightCountryIndex === index
            });

            var inputFlagClasses = 'flag ' + country.iso2;

            return React.createElement(
                'li',
                {
                    ref: 'flag_no_' + index,
                    key: 'flag_no_' + index,
                    'data-flag-key': 'flag_no_' + index,
                    className: itemClasses,
                    'data-dial-code': '1',
                    'data-country-code': country.iso2,
                    onClick: self.handleFlagItemClick.bind(self, country) },
                React.createElement('div', { className: inputFlagClasses, style: self.getFlagStyle() }),
                React.createElement(
                    'span',
                    { className: 'country-name' },
                    country.name
                ),
                React.createElement(
                    'span',
                    { className: 'dial-code' },
                    '+' + country.dialCode
                )
            );
        });

        var dashedLi = React.createElement('li', { key: "dashes", className: 'divider' });
        // let's insert a dashed line in between preffered countries and the rest
        countryDropDownList.splice(this.state.preferredCountries.length, 0, dashedLi);

        var dropDownClasses = classNames({
            'country-list': true,
            'hide': !this.state.showDropDown
        });
        return React.createElement(
            'ul',
            { ref: 'flagDropdownList', className: dropDownClasses },
            countryDropDownList
        );
    },
    getFlagStyle: function getFlagStyle() {
        return {
            width: 16,
            height: 11,
            backgroundImage: 'url(' + this.props.flagsImagePath + ')'
        };
    },
    handleInputBlur: function handleInputBlur() {
        if (typeof this.props.onBlur === 'function') {
            this.props.onBlur(this.state.formattedNumber, this.state.selectedCountry);
        }
    },
    render: function render() {
        var arrowClasses = classNames({
            'arrow': true,
            'up': this.state.showDropDown
        });
        var inputClasses = classNames({
            'form-control': true,
            'invalid-number': !this.props.isValid(this.state.formattedNumber.replace(/\D/g, ''))
        });

        var flagViewClasses = classNames({
            'flag-dropdown': true,
            'open-dropdown': this.state.showDropDown
        });

        var inputFlagClasses = 'flag ' + this.state.selectedCountry.iso2;
        var otherProps = {};
        if (this.props.inputId) {
            otherProps.id = this.props.inputId;
        }
        return React.createElement(
            'div',
            { className: classNames('react-tel-input', this.props.classNames, this.props.className) },
            React.createElement('input', _extends({
                onChange: this.handleInput,
                onClick: this.handleInputClick,
                onFocus: this.handleInputFocus,
                onBlur: this.handleInputBlur,
                onKeyDown: this.handleInputKeyDown,
                value: this.state.formattedNumber,
                ref: 'numberInput',
                type: 'tel',
                className: inputClasses,
                autoComplete: this.props.autoComplete,
                pattern: this.props.pattern,
                required: this.props.required,
                placeholder: this.props.placeholder,
                disabled: this.props.disabled }, otherProps)),
            React.createElement(
                'div',
                { ref: 'flagDropDownButton', className: flagViewClasses, onKeyDown: this.handleKeydown },
                React.createElement(
                    'div',
                    { ref: 'selectedFlag', onClick: this.handleFlagDropdownClick, className: 'selected-flag', title: this.state.selectedCountry.name + ': + ' + this.state.selectedCountry.dialCode },
                    React.createElement(
                        'div',
                        { className: inputFlagClasses, style: this.getFlagStyle() },
                        React.createElement('div', { className: arrowClasses })
                    )
                ),
                this.state.showDropDown ? this.getCountryDropDownList() : ''
            )
        );
    }
});

exports.default = onClickOutside(ReactTelephoneInput);
