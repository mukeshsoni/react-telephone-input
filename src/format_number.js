import R from 'cramda'

const { first, tail } = R

function formatNumber(text, pattern, autoFormat, length) {
  if (!text || text.length === 0) {
    return '+'
  }

  // for all strings with length less than 3, just return it (1, 2 etc.)
  // also return the same text if the selected country has no fixed format
  if ((text && text.length < 2) || !pattern || !autoFormat) {
    return `+${text}`
  }

  if (length === null) {
    length = text.length;
  }

  if (text.length > length) {
    text = text.substring(0, length-1);
  }

  const formattedObject = pattern.split('').reduce(
    (acc, character) => {
      if (acc.remainingText.length === 0) {
        return acc
      }

      if (character !== '.') {
        return {
          formattedText: acc.formattedText + character,
          remainingText: acc.remainingText,
        }
      }

      return {
        formattedText: acc.formattedText + first(acc.remainingText),
        remainingText: tail(acc.remainingText),
      }
    },
    { formattedText: '', remainingText: text.split('') },
  )
  return formattedObject.formattedText + formattedObject.remainingText.join('')
}

export default formatNumber
