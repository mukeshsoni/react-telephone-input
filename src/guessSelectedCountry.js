// memoize results based on the first 5/6 characters. That is all that matters

import R from 'cramda';
import countryData from 'country-telephone-data';

const { find, propEq, startsWith } = R;
const { allCountries, allCountryCodes } = countryData;

export default function guessSelectedCountry(inputNumber, props) {
  const { defaultCountry, onlyCountries } = props;

  const secondBestGuess =
    find(propEq('iso2', defaultCountry), allCountries) || onlyCountries[0];

  const inputNumberForCountries = inputNumber.substr(0, 4);
  let bestGuess;

  if (inputNumber.trim() !== '') {
    bestGuess = onlyCountries.reduce(
      (selectedCountry, country) => {
        // if the country dialCode exists WITH area code

        if (
          allCountryCodes[inputNumberForCountries] &&
          allCountryCodes[inputNumberForCountries][0] === country.iso2
        ) {
          return country;

          // if the selected country dialCode is there with the area code
        } else if (
          allCountryCodes[inputNumberForCountries] &&
          allCountryCodes[inputNumberForCountries][0] === selectedCountry.iso2
        ) {
          return selectedCountry;

          // else do the original if statement
        }
        if (startsWith(country.dialCode, inputNumber)) {
          if (country.dialCode.length > selectedCountry.dialCode.length) {
            return country;
          }
          if (
            country.dialCode.length === selectedCountry.dialCode.length &&
            country.priority < selectedCountry.priority
          ) {
            return country;
          }
        }

        return selectedCountry;
      },
      { dialCode: '', priority: 10001 },
      this,
    );
  } else {
    return secondBestGuess;
  }

  if (!bestGuess || !bestGuess.name) {
    return secondBestGuess;
  }

  return bestGuess;
}
