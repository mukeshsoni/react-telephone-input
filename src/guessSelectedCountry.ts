import * as React from 'react';
// memoize results based on the first 5/6 characters. That is all that matters

import R from 'cramda';
import countryData from 'country-telephone-data';

const { find, propEq, startsWith } = R;
const { allCountries, allCountryCodes } = countryData;
import { Country, Props } from './ReactTelephoneInput';

export default function guessSelectedCountry(
  inputNumber: string,
  props: Props,
): Country {
  const defaultCountry = props.defaultCountry!;
  const onlyCountries = props.onlyCountries!;

  const secondBestGuess =
    find(propEq('iso2', defaultCountry), allCountries) || onlyCountries[0];

  const inputNumberForCountries = inputNumber.substr(0, 4);
  let bestGuess;

  if (inputNumber.trim() !== '') {
    bestGuess = onlyCountries.reduce(
      (selectedCountry: Country, country: Country) => {
        // if the country dialCode exists WITH area code

        if (
          allCountryCodes[inputNumberForCountries] &&
          allCountryCodes[inputNumberForCountries][0] === country.iso2
        ) {
          return country;

          // if the selected country dialCode is there with the area code
        } else if (
          allCountryCodes[inputNumberForCountries] &&
          allCountryCodes[inputNumberForCountries][0] === selectedCountry.iso2
        ) {
          return selectedCountry;

          // else do the original if statement
        }
        if (startsWith(country.dialCode, inputNumber)) {
          if (country.dialCode.length > selectedCountry.dialCode.length) {
            return country;
          }
          if (
            country.dialCode.length === selectedCountry.dialCode.length &&
            country.priority < selectedCountry.priority
          ) {
            return country;
          }
        }

        return selectedCountry;
      },
      { dialCode: '', priority: 10001 },
    );
  } else {
    return secondBestGuess;
  }

  if (!bestGuess || !bestGuess.name) {
    return secondBestGuess;
  }

  return bestGuess;
}
