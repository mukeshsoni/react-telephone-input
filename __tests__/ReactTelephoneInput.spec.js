import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import countryData from 'country-telephone-data';

import { ReactTelephoneInput } from '../src/ReactTelephoneInput';
import guessSelectedCountry from '../src/guessSelectedCountry';

const { allCountries } = countryData;

expect.extend({
  toBeType(received, argument) {
    const initialType = typeof received;
    const type =
      initialType === 'object'
        ? Array.isArray(received)
          ? 'array'
          : initialType
        : initialType;
    return type === argument
      ? {
          message: () => `expected ${received} to be type ${argument}`,
          pass: true,
        }
      : {
          message: () => `expected ${received} to be type ${argument}`,
          pass: false,
        };
  },
});

describe('react telephone input', () => {
  it('mandatory existential crisis test', () => {
    const { container } = render(<ReactTelephoneInput />);
    expect(container).toBeDefined();
  });

  it('should render the top divs and inputses', () => {
    const { container } = render(<ReactTelephoneInput />);

    expect(container.querySelectorAll('div.react-tel-input')).toHaveLength(1);
    expect(container.querySelectorAll('input')).toHaveLength(1);
  });

  it('should show the placeholder as passed in the prop', () => {
    const placeholder = '0001-121-212';
    const { queryByPlaceholderText } = render(
      <ReactTelephoneInput placeholder={placeholder} />,
    );

    expect(queryByPlaceholderText(placeholder)).toBeDefined();
  });

  it('should call onEnterKeyPress prop callback on enter key press', () => {
    const onEnterKeyPress = jest.fn();
    const { container } = render(
      <ReactTelephoneInput
        defaultCountry="us"
        initialValue="+9112121"
        preferredCountries={['us', 'ca', 'zz', 'hk']}
        onEnterKeyPress={() => {
          onEnterKeyPress();
        }}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
    const input = container.querySelector('input');
    input.focus();
    container.querySelector('input').focus();
    fireEvent.change(input, {
      target: { value: '+91 12121-1212', checked: false },
    });
    fireEvent.keyDown(input, {
      key: 'Enter',
      code: 'Enter',
      which: 13,
      keyCode: 13,
    });
    expect(onEnterKeyPress.mock.calls.length).toBe(1);
    expect(container.firstChild).toMatchSnapshot();
  });

  // refer issue - https://github.com/mukeshsoni/react-telephone-input/issues/103
  it('should guess correct country when flag is changed manually from the dropdown', () => {
    const { container } = render(
      <ReactTelephoneInput
        defaultCountry="us"
        preferredCountries={['us', 'ca', 'zz', 'hk']}
      />,
    );

    expect(container.querySelectorAll('div.flag-dropdown')).toHaveLength(1);
    expect(container.querySelectorAll('.selected-flag > div.us')).toHaveLength(
      1,
    );

    // the dropdown list is not there
    expect(container.querySelector('.country-list')).toBeNull();
    // let's click on the selected flag
    fireEvent.click(container.querySelector('.selected-flag'));
    expect(container.querySelector('div.country-list')).not.toBeNull();

    const countryList = container.querySelectorAll(
      'div.country-list > div > div',
    );
    // the preferred list should put canada in the second position in country
    // dropdown
    // now let's click on canada
    fireEvent.click(countryList[1]);
    expect(container.querySelectorAll('div.country-list')).toHaveLength(0);
    expect(container.querySelectorAll('.selected-flag div.ca')).toHaveLength(1);
  });

  it('should allow custom value for autoComplete input property', () => {
    const { rerender, container } = render(<ReactTelephoneInput />);
    expect(
      container.querySelector('input').getAttribute('autocomplete'),
    ).toEqual('tel');

    rerender(<ReactTelephoneInput autoComplete="off" />);
    expect(
      container.querySelector('input').getAttribute('autocomplete'),
    ).toEqual('off');
  });

  it('should guess selected country', () => {
    const props = {
      onlyCountries: allCountries,
      defaultCountry: allCountries[0].iso2,
    };

    // if nothing is sent in, select the first country in allCountries list as the default
    expect(guessSelectedCountry('', props).iso2).toEqual(allCountries[0].iso2);

    // if input value is sent, select appropriately
    expect(guessSelectedCountry('12', props).iso2).toEqual('us'); // based on priority
    expect(guessSelectedCountry('12112121', props).iso2).toEqual('us');
    expect(guessSelectedCountry('913212121', props).iso2).toEqual('in');
    expect(guessSelectedCountry('237', props).iso2).toEqual('cm'); // based on priority
    expect(guessSelectedCountry('599', props).iso2).toEqual('cw');
    expect(guessSelectedCountry('590', props).iso2).toEqual('gp');
    expect(guessSelectedCountry('1403', props).iso2).toEqual('ca');
    expect(guessSelectedCountry('18005', props).iso2).toEqual('us');
    expect(guessSelectedCountry('1809', props).iso2).toEqual('do');

    // select the first one if not able to resolve completely
    expect(guessSelectedCountry('59', props).iso2).toEqual(
      allCountries[0].iso2,
    );
  });

  it('should set the correct highlightCountryIndex', () => {
    const afghanistan = {
      name: 'Afghanistan (‫افغانستان‬‎)',
      iso2: 'af',
      dialCode: '93',
      priority: 0,
    };
    const albania = {
      name: 'Albania (Shqipëri)',
      iso2: 'al',
      dialCode: '355',
      priority: 0,
    };
    const algeria = {
      name: 'Algeria (‫الجزائر‬‎)',
      iso2: 'dz',
      dialCode: '213',
      priority: 0,
      format: '+...-..-...-....',
    };

    // Setup ReactTelephoneInput with a fixed set of countries
    // so we know the expected indexes for sure
    const { container } = render(
      <ReactTelephoneInput
        onlyCountries={[afghanistan, albania, algeria]}
        preferredCountries={[algeria.iso2]}
        initialValue="+121345"
      />,
    );

    fireEvent.click(container.querySelector('.selected-flag'));

    let countryList = container.querySelectorAll(
      'div.country-list > div > div',
    );
    fireEvent.click(countryList[0]);
    let input = container.querySelector('input');
    expect(input.value).toEqual('+213-12-134-5');

    fireEvent.click(container.querySelector('.selected-flag'));
    countryList = container.querySelectorAll('div.country-list > div > div');
    fireEvent.click(countryList[1]);
    input = container.querySelector('input');
    expect(input.value).toEqual('+93121345');

    fireEvent.click(container.querySelector('.selected-flag'));
    countryList = container.querySelectorAll('div.country-list > div > div');
    fireEvent.click(countryList[2]);
    input = container.querySelector('input');
    expect(input.value).toEqual('+355121345');
  });

  it('should trigger onFocus event handler when input element is focused', (done) => {
    const onFocus = (number, country) => {
      expect(number).toBeType('string');
      expect(country).toBeType('object');
      done();
    };

    const { container } = render(<ReactTelephoneInput onFocus={onFocus} />);

    fireEvent.focus(container.querySelector('input'));
  });

  it('should trigger onBlur event handler when input element is unfocused', (done) => {
    const onBlur = (number, country) => {
      expect(number).toBeType('string');
      expect(country).toBeType('object');
      done();
    };

    const { container } = render(<ReactTelephoneInput onBlur={onBlur} />);

    fireEvent.blur(container.querySelector('input'));
  });

  it('should re-render with correct phone number once value prop changed', () => {
    const { container, rerender } = render(
      <ReactTelephoneInput value="+12313123132" />,
    );
    let input = container.querySelector('input');
    expect(input.value).toEqual('+1 (231) 312-3132');
    rerender(<ReactTelephoneInput value="+12313123133" />);
    input = container.querySelector('input');
    expect(input.value).toEqual('+1 (231) 312-3133');
  });
});
