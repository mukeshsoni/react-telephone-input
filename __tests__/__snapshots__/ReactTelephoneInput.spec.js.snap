// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react telephone input should call onEnterKeyPress prop callback on enter key press 1`] = `
<div
  class="react-tel-input"
  data-test-id="src_reacttelephoneinput_test_id_4"
>
  <div
    class="flag-dropdown"
    data-test-id="src_reacttelephoneinput_test_id_6"
  >
    <button
      class="selected-flag"
      data-test-id="src_reacttelephoneinput_test_id_7"
      title="India (भारत): + 91"
      type="button"
    >
      <div
        class="flag in"
        data-test-id="src_reacttelephoneinput_test_id_8"
        style="background-image: url(flags.png);"
      >
        <div
          class="arrow"
          data-test-id="src_reacttelephoneinput_test_id_9"
        />
      </div>
    </button>
    
  </div>
  <input
    autocomplete="tel"
    class="form-control"
    data-test-id="src_reacttelephoneinput_test_id_5"
    placeholder="+1 (702) 123-4567"
    type="tel"
    value="+91 12121"
  />
</div>
`;

exports[`react telephone input should call onEnterKeyPress prop callback on enter key press 2`] = `
<div
  class="react-tel-input"
  data-test-id="src_reacttelephoneinput_test_id_4"
>
  <div
    class="flag-dropdown"
    data-test-id="src_reacttelephoneinput_test_id_6"
  >
    <button
      class="selected-flag"
      data-test-id="src_reacttelephoneinput_test_id_7"
      title="India (भारत): + 91"
      type="button"
    >
      <div
        class="flag in"
        data-test-id="src_reacttelephoneinput_test_id_8"
        style="background-image: url(flags.png);"
      >
        <div
          class="arrow"
          data-test-id="src_reacttelephoneinput_test_id_9"
        />
      </div>
    </button>
    
  </div>
  <input
    autocomplete="tel"
    class="form-control"
    data-test-id="src_reacttelephoneinput_test_id_5"
    placeholder="+1 (702) 123-4567"
    type="tel"
    value="+91 12121-1212"
  />
</div>
`;
